server/aesdsocket.c

```diff
@@ -20,82 +20,581 @@
+        new_node = NULL;
+
+        // check for thread completion
+        SLIST_FOREACH(new_node, &head, nodes)
+        {
+            if(new_node->thread_data.thread_complete)
+            {
+                pt_ret = pthread_join(new_node->thread_data.thread_id,&thread_rtn);
+                if(pt_ret == RET_ERROR)
+                {
+                    syslog(LOG_ERR, "Thread join failed");
+                    return -1;
+                }
+                if(thread_rtn == NULL)
+                {
+                    return -1;
+                }
+                syslog(LOG_INFO, "Thread join %ld",new_node->thread_data.thread_id);
+            }
+        } 
+    }
+
+    return 0;
+}
+
+// thread function for receive and send commmands
+void *recv_send_thread(void *thread_param)
+{
+    int ret;
+    // receive bytes
+	ssize_t recv_bytes = 0;
+	char recv_buf[BUF_LEN];
+	
+	// send bytes
+	ssize_t send_bytes = 0;
+	char send_buf[BUF_LEN];
+	ssize_t bytes_read;
+
+    // to print IP
+    char s[INET6_ADDRSTRLEN];
+
+    memset(recv_buf, 0, BUF_LEN);
+    memset(send_buf, 0, BUF_LEN);
+
+    thread_data_t *thread_data = (thread_data_t*)thread_param;
+
+    inet_ntop(thread_data->client_addr->ss_family,
+                get_in_addr((struct sockaddr *)&(thread_data->client_addr)),
+                s, sizeof s);
+    syslog(LOG_INFO,"Accepted connection from %s",s);
+
+    syslog(LOG_INFO,"Started thread %ld",thread_data->thread_id);
+
+    /********************************************************* 
+    *  STEP 4 : 
+    *  Receives data over the connection and 
+    *  appends to file /var/tmp/aesdsocketdata
+    *********************************************************/
+
+    // acquire lock
+    ret = pthread_mutex_lock(thread_data->mutex);
+    if(ret == RET_ERROR)
+    {
+        syslog(LOG_ERR,"mutex lock failed\n");
+        return NULL;
+    }
+
+    // receive and write
+    do
+    {
+        // receive data on socket
+        recv_bytes = recv(thread_data->accept_fd, recv_buf, BUF_LEN, 0);
+        if(recv_bytes == RET_ERROR)
+        {
+            syslog(LOG_ERR,"Receive failed");
+            return NULL;
+        }
+
+        // write data to file
+        ret = write(data_file_fd, recv_buf, recv_bytes);
+        if(ret == RET_ERROR)
+        {
+            syslog(LOG_ERR,"File write failed");
+            return NULL;
+        }
+    }while((memchr(recv_buf, '\n', recv_bytes)) == NULL);
+
+    // release lock
+    ret = pthread_mutex_unlock(thread_data->mutex);
+    if(ret == RET_ERROR)
+    {
+        syslog(LOG_ERR,"mutex unlock failed\n");
+        return NULL;
+    }
+
+    /********************************************************* 
+    *  STEP 5 : 
+    *  Returns the full content of /var/tmp/aesdsocketdata 
+    *  to the client as soon as the received data packet 
+    *  completes.
+    *********************************************************/
+    // acquire lock
+    ret = pthread_mutex_lock(thread_data->mutex);
```

Comment:
@ameyflash think about minimum locking interval.  As implemented there's not much multiple threads can do since most of the function is locked